### PR TITLE
Add properties of `Data.Refinement` and refactor module structure

### PR DESCRIPTION
**Pull request by jamesmckinna**
NB.
* could be folded into `Data.Refinement` directly?
* Fairbairn threshold? is this injectivity proof not *immediate*?

### CODE REVIEW

File: src/Data/Refinement/Properties.agda [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+open import Level
+open import Data.Refinement
+open import Relation.Binary.PropositionalEquality.Core using (_≡_; refl)
+
+private
+  variable
+    a p : Level
+    A : Set a
+
+module _ {P : A → Set p} where
```
Comment (MatthewDaggitt) [inline src/Data/Refinement/Properties.agda:20]: As `P` is used in the type of `Refinement`, it's fine to add it as a variable. In fact you could do this in `Data.Refinement` as well if you were feeling keen.

**Review by MatthewDaggitt — APPROVED** [review]
Yup, I this is definitely useful! Also I think `Properties` is definitely the right place for it, even though it seems like overkill at the moment. I'm sure more stuff will be added to it!

**Review by gallais — APPROVED** [review]

**Comment by jamesmckinna** [comment]
> Yup, I this is definitely useful! Also I think `Properties` is definitely the right place for it, even though it seems like overkill at the moment. I'm sure more stuff will be added to it!

Two things: 
* I do have some additional properties (concerning iterated refinement) which occurred to me during the writing of this (downstream PR?)
* Should we followup with a refactoring into an orthodox `.Base`/`.Properties` division? (also could be handled downstream)

File: src/Data/Refinement/Properties.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Data/Refinement/Properties.agda:20]: Will do!

**Comment by gallais** [comment]
Yes please: if it grows enough, let's split it up according to the hierarchy!

**Comment by jamesmckinna** [comment]
OK, I've added a proof of (inheritance of) decidable equality, and refactored as `Base`/`Properties`. I'll stop there.

**Comment by jamesmckinna** [comment]
Does @fredrikNordvallForsberg 's resolution of #2495 mean that we could simplify the implementation of `Data.Refinement` to simply make the `proof` field take the `.` modality?
```agda
record Refinement {a p} (A : Set a) (P : A → Set p) : Set (a ⊔ p) where
  constructor _,_
  field value : A
        .proof : P value
infixr 4 _,_
open Refinement public

module _ {P : A → Set p} {Q : B → Set q} where

  map : (f : A → B) → ∀[ P ⇒ f ⊢ Q ] →
        [ a ∈ A ∣ P a ] → [ b ∈ B ∣ Q b ]
  map f prf (a , p) = f a , prf p
```
etc.

Not for this PR, but a potential simplifying refactoring/deprecation downstream?

**Comment by gallais** [comment]
You don't get irrelevant projections so IME it's nicer to package up things affected by the irrelevant
modality so that you can project them out and pass them around to combinators.

**Comment by JacquesCarette** [comment]
The old category library used to use irrelevance *a lot* and then broke in ways we couldn't fix at some point in Agda's evolution. So I'd be quite concerned that we'd run up against bugs "just too late" if we start using it too much in the library.

Irrelevant projections (and the lack thereof) are a big deal. Conor and I more recently tried to use irrelevance quite aggressively to force Agda to ensure we stay honest (we're doing quotients "by hand"), and gave up. Now we stay honest by sheer force of will.

**Comment by jamesmckinna** [comment]
Ready to merge now?

File: src/Data/Refinement/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+-- Basic properties
+
+value-injective : value v ≡ value w → v ≡ w
+value-injective refl = refl
+
+module _ (eq? : DecidableEquality A) where
+
+  infix 4 _≟_
+  _≟_ : DecidableEquality [ x ∈ A ∣ P x ]
+  v ≟ w = Dec.map′ value-injective (cong value) (eq? (value v) (value w))
```
Comment (jamesmckinna) [inline src/Data/Refinement/Properties.agda:35]: This is clearly a pattern, which currently is reified only under `Relation.Decidable.Nullary.Decidable.via-injection`, but it seems excessive to create an `Injection` bundle solely for the sake of applying the lemma, so suggest some downstream refactoring to make this a property of `Injective` (bare) functions?

**Review by JacquesCarette — APPROVED** [review]

**Pull request merged**